standard/statements.md

```diff
@@ -1308,6 +1314,80 @@ The order in which `foreach` traverses the elements of an array, is as follows:
 >
 > *end example*
 
+#### §async-foreach await foreach
+
+The compile-time processing of a `foreach` statement first determines the ***collection type***, ***enumerator type*** and ***iteration type*** of the expression. The processing for a `foreach` statement is detailed in §sync-foreach and the process for an `await foreach` is detailed in §async-foreach.
+
+This determination proceeds as follows:
+
+- Determine whether the type `X` has an appropriate `GetAsyncEnumerator` method:
+  - Perform member lookup on the type `X` with identifier `GetAsyncEnumerator` and no type arguments. If the member lookup does not produce a match, or it produces an ambiguity, or produces a match that is not a method group, check for an enumerable interface as described below. It is recommended that a warning be issued if member lookup produces anything except a method group or no match.
+  - Perform overload resolution using the resulting method group and an empty argument list. If overload resolution results in no applicable methods, results in an ambiguity, or results in a single best method but that method is either static or not public, check for an enumerable interface as described below. It is recommended that a warning be issued if overload resolution produces anything except an unambiguous public instance method or no applicable methods.
+  - If the return type `E` of the `GetAsyncEnumerator` method is not a class, struct or interface type, an error is produced and no further steps are taken.
+  - Member lookup is performed on `E` with the identifier `Current` and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a public instance property that permits reading, an error is produced and no further steps are taken.
+  - Member lookup is performed on `E` with the identifier `MoveNextAsync` and no type arguments. If the member lookup produces no match, the result is an error, or the result is anything except a method group, an error is produced and no further steps are taken.
```

Comment:
In both of these bullets, we have two different error conditions, but we describe them separately: "If X happens, an error is produced; if Y happens, an error is produced". Can't we have "If X or Y happens, an error is produced"?

Reply:
I tried that rewrite, and the language was clumsier because one is a method group and the other is a property. Leaving as is.